### PR TITLE
chore: dyntypes decoding improvements

### PR DESCRIPTION
**Pull request by dpc**

### CODE REVIEW

File: fedimint-client/src/lib.rs [hunk 1]
```diff
@@ -1758,7 +1758,7 @@ impl ClientBuilder {
             module_inits: client.module_inits.clone(),
             primary_module_instance: Some(client.primary_module_instance),
             admin_creds: None,
-            db: client.db.clone(),
+            db_no_encoders: client.db.with_decoders(Default::default()),
```
Comment (dpc) [inline fedimint-client/src/lib.rs:1761]: Notice - in `ClientBuilder`, the `db` should not contain any built-in decoders. It's somewhat important, and in here e.g. we would unnecessarily inherit decoders from previous `Client`, which is harmless right now, but could lead to bugs. For this reason I decided to rename it to this somewhat odd looking name.

File: fedimint-core/src/core.rs [hunk 2]
```diff
@@ -322,15 +343,53 @@ impl Decoder {
         DecoderBuilder::default()
     }
 
+    /// System Dyn-type, don't use.
+    #[doc(hidden)]
+    pub fn builder_system() -> DecoderBuilder {
```
Comment (dpc) [inline fedimint-core/src/core.rs:348]: A bit of a ... workaround... . `DynState` is not really a `DynType` in the same sense as `DynInput` etc. are - it doesn't belong to a real module. But it does work exactly the same, so it's convenient to just make up and ID etc. for it. But architecturally it's a hack, and it's going to lead to papercuts and workarounds like here.

File: fedimint-core/src/core.rs [hunk 3]
```diff
@@ -322,15 +343,53 @@ impl Decoder {
...
+        &self,
+        reader: &mut dyn Read,
+        total_len: u64,
+        module_id: ModuleInstanceId,
+        decoders: &ModuleDecoderRegistry,
+    ) -> Result<DynType, DecodeError> {
+        let mut reader = reader.take(total_len);
+
+        let val = self.decode_partial(&mut reader, module_id, decoders)?;
+        let left = reader.limit();
```
Comment (dpc) [inline fedimint-core/src/core.rs:369]: Originally with @maan2003 when debugging, we were printing the bytes here as well, but I decided against it as it opens the code to problems: e.g. what if reading leftover bytes leads to IO errors? Which error we return, which ignore. Also - what if the reader blocks, etc. Seems not worth it.

Comment (maan2003) [reply fedimint-core/src/core.rs:369]: just first print error without bytes, then read bytes and print bytes in seperate message. if all this work, return the error to caller

File: fedimint-core/src/core.rs (hunk 2 above)
Comment (elsirion) [reply fedimint-core/src/core.rs:348]: I think we could get rid of nested dyn types by serializing the transaction before putting it into the tx state machine. Over the wire it's sent as hex-encoded bytes anyway (using `SerdeModuleEncoding<Transaction>`).

File: fedimint-client/src/lib.rs (hunk 1 above)
Comment (elsirion) [reply fedimint-client/src/lib.rs:1761]: Now that the client can shut down cleanly we should get rid of `from_existing`, that was our old workaround to not have to re-open the DB.

File: fedimint-client/src/lib.rs [hunk 4]
```diff
@@ -1822,12 +1822,12 @@ impl ClientBuilder {
         Ok(())
     }
 
-    pub fn db(&self) -> &Database {
-        &self.db
+    pub fn db_no_encoders(&self) -> &Database {
```
Comment (elsirion) [inline fedimint-client/src/lib.rs:1825]: Should be called `db_no_decoders` imo, encoding doesn't require any registry, only decoding.

File: fedimint-core/src/core.rs (hunk 2 above)
Comment (dpc) [reply fedimint-core/src/core.rs:348]: Might require migration, possibly extra work -> outside of scope of this change. :)

File: fedimint-core/src/core.rs (hunk 3 above)
Comment (dpc) [reply fedimint-core/src/core.rs:369]: I still don't like it. :D

File: fedimint-client/src/lib.rs (hunk 4 above)
Comment (dpc) [reply fedimint-client/src/lib.rs:1825]: Oops. brainfart.

**Review by elsirion — DISMISSED** [review]

**Review by elsirion — APPROVED** [review]
I trust you didn't make major changes since last review, at least it didn't look like it.

**Review by maan2003 — APPROVED** [review]

**Pull request merged**